fix(exception_handling): re-raising java exception caused an error

jimtng opened the pull request:
Resolve  #406

ccutrer commented on lib/openhab/log/logger.rb:
```diff
@@ -79,7 +79,7 @@ def clean_backtrace(error)
           error.set_stack_trace(backtrace)
         end
-        error
+        error.is_a?(Exception) ? error : JavaException.new(error.message)
```
I'd recommend passing the original error to JavaException.new, and then in an initialize method in that class store the entire error to an ivar, and only pass the message to super. You should also look if jruby already has a similar Ruby class defined

jimtng replied:
Thanks for your suggestion. What should be done with that ivar?

The first problem is that I don't know how to reproduce this issue. So far I very rarely (but not never) encountered it on my production system. However, I see this error most often in the cucumber tests on the github workflow.

Also upon further investigation, this `error` came from `StandardError => e`

https://github.com/boc-tothefuture/openhab-jruby/blob/409c7a07586dc0a2cdf828317a72790a90ccf51a/lib/openhab/dsl/rules/rule.rb#L51

So it seems that it should've already been an instance of StandardError, but why re-raising it can cause 
```
(TypeError) exception class/object expected
```

ccutrer replied:
Just accessible with an attr_reader. As for how it could happen... no idea.

ccutrer replied:
Maybe add a rescue around the raise that if it does a TypeError, log the details of what you're trying to raise and then re-raise (the TypeError).

Perhaps there's a funny jruby thing going on that StandardError#=== is catching something Java-ish, but internally it can't re-raise

jimtng replied:
Who is going to consume / use that attribute? Do I just pick an arbitrary name, e.g. `exception_object` or something?

ccutrer replied:
Anyone interested in the original error. On second thought, there might be an implicit `cause` accessor that's populated. Please to double check.

jimtng replied:
> Maybe add a rescue around the raise that if it does a TypeError, log the details of what you're trying to raise and then re-raise (the TypeError).

That's actually a brilliant idea!

jimtng replied:
I've removed the JavaException class for now and just log the error that caused the TypeError as you suggested.

jimtng commented:
Note: need to change the commit message to something like:
`refactor(exception_handling): log the exception that caused TypeError`

ccutrer reviewed (changes requested):
Ah. Yes, this is not the right way. https://github.com/ccutrer/homie-mqtt/blob/6fcfe51e76bc23494c6f67ce5e28422b3a9087a4/lib/mqtt/homie/device.rb#L140 is the proper way to change the backtrace of an existing exception then re-raise it.

jimtng commented:
Yes, it was calling `raise e, message` with e an exception object, not a class, hence TypeError.

jimtng commented:
I figured out how to trigger this error: An error that occurred during rule creation, not execution. That's the only place re_raise_with_backtrace is called. I've now added a cucumber test for this scenario.

Upon further investigation, other places that called clean_backtrace don't re-raise the exception - I suppose due to the difficulties we're facing here.

So I opted to do the same, and use print_backtrace instead, which is now moved to logger to DRY it.

I am still unsure whether print_backtrace (which calls logger.error) is the ideal way to do it, or whether it should be re-raised. However, I'm happy with this for now unless anyone has objections.

jimtng commented:
I renamed `print_backtrace` to `log_exception`. This is ready to be reviewed / merged.
